<a href="https://colab.research.google.com/github/maykends/Processamento-de-Linguagem-Natural/blob/master/Twitter_Rede_Neural_Recorrente.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Classificações de Sentimentos do Twitter usando RNN

**Classificação de  sentimentos do Twitter.**

Neste notebook iremos treinar uma Rede Neural Recorrente com o intuito de classificar sentimentos do *Twitter em uma RNN*.

- **Aluno(a):** Mayke Kend Gomes Delgado.
- **Prof.** Jones Granatier
- **Curso:** Pós-Graduação em Ciência de Dados.
- **Disciplina** **169025** - Processamento de Linguagem Natural
- **Matrícula:** 1691690020

## Etapa 1: Instalando o TensorFlow 2.0

In [ ]:
!pip uninstall -y tensorflow #Comando necessário, pois o TensorFlow-gpu não desinstala a versão mais recente do Tensorflow, pode gerar conflitos.

In [ ]:
!pip install tensorflow-gpu==2.0

In [ ]:
!pip install numpy==1.17

## Etapa 2: Importação das bibliotecas

In [ ]:
import tensorflow as tf
import numpy as np

In [ ]:
tf.__version__

'2.0.0'

## Etapa 3: Pré-processamento

### Configurando os parâmetros para a base de dados

In [ ]:
# número de palavras, das quais são mais frequentes para se utilizar
number_of_words = 20000 # 170.000 -> 20.000/40.000 -> 3.000
# cortar os textos a partir do maxlenght = 100 mediante ao seu comprimento máximo de palavras
max_len = 500 # 500

### Carregando a base de dados IMDB

In [ ]:
# (X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=number_of_words)
# train_inputs, test_inputs, train_labels, test_labels

import pickle
with open('save.pkl', 'rb') as f:
  X_train, X_test, y_train, y_test = pickle.load(f)



In [ ]:
X_train.shape

(2450, 60)

In [ ]:
X_train # EOF - end of sentence, 1 = SOS (start of sentence)

array([[ 393,  108,   33, ...,    0,    0,    0],
       [   2,    4,  233, ...,    0,    0,    0],
       [   2, 1865, 2810, ...,    0,    0,    0],
       ...,
       [   2,  118,  424, ...,    0,    0,    0],
       [   2,   15,  318, ...,    0,    0,    0],
       [ 106,   17,  106, ...,    0,    0,    0]], dtype=int32)

In [ ]:
X_train[0]

array([ 393,  108,   33,   89, 6185,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0], dtype=int32)

In [ ]:
y_train

array([0, 1, 0, ..., 1, 0, 1])

### Preenchimento das sequências (textos) para terem o mesmo tamanho

In [ ]:
len(X_train[0])

60

In [ ]:
len(X_train[1])

60

In [ ]:
X_train = tf.keras.preprocessing.sequence.pad_sequences(X_train, maxlen=max_len)

In [ ]:
len(X_train[0])

500

In [ ]:
len(X_train[1])

500

In [ ]:
X_test = tf.keras.preprocessing.sequence.pad_sequences(X_test, maxlen=max_len)

## Etapa 4: Construindo a Rede Neural Recorrente

### Definindo o modelo

In [ ]:
model = tf.keras.Sequential()

### Adicionando a camada de embedding

In [ ]:
X_train.shape[1]

500

In [ ]:
# output_dim = 256

model.add(tf.keras.layers.Embedding(input_dim=number_of_words, output_dim=256, input_shape=(X_train.shape[1],)))

### Adicionando a camada LSTM

- units: 128
- activation: tanh

In [ ]:
model.add(tf.keras.layers.LSTM(units=32, activation='tanh'))

### Adicionando a camada de saída

- units: 1
- activation: sigmoid

In [ ]:
model.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

### Compilando o modelo

In [ ]:
model.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['accuracy'])

In [ ]:
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_2 (Embedding)      (None, 500, 256)          5120000   
_________________________________________________________________
lstm_2 (LSTM)                (None, 32)                36992     
_________________________________________________________________
dense_2 (Dense)              (None, 1)                 33        
Total params: 5,157,025
Trainable params: 5,157,025
Non-trainable params: 0
_________________________________________________________________


### Treinando o modelo

In [ ]:
# treinando nosso modelo com 10 épocas
model.fit(X_train, y_train, epochs=20, batch_size=128)

Train on 2450 samples
Epoch 1/20
2450/2450 [==============================] - 2s 706us/sample - loss: 0.6630 - accuracy: 0.5673
Epoch 2/20
2450/2450 [==============================] - 2s 663us/sample - loss: 0.6537 - accuracy: 0.6245
Epoch 3/20
2450/2450 [==============================] - 2s 655us/sample - loss: 0.5835 - accuracy: 0.7465
Epoch 4/20
2450/2450 [==============================] - 2s 655us/sample - loss: 0.6427 - accuracy: 0.6196
Epoch 5/20
2450/2450 [==============================] - 2s 646us/sample - loss: 0.5706 - accuracy: 0.7412
Epoch 6/20
2450/2450 [==============================] - 2s 646us/sample - loss: 0.5492 - accuracy: 0.8045
Epoch 7/20
2450/2450 [==============================] - 2s 633us/sample - loss: 0.4394 - accuracy: 0.8600
Epoch 8/20
2450/2450 [==============================] - 2s 632us/sample - loss: 0.3720 - accuracy: 0.8792
Epoch 9/20
2450/2450 [==============================] - 2s 641us/sample - loss: 0.4006 - accuracy: 0.8457
Epoch 10/20
2450/2450 [=

### Avaliando o modelo

In [ ]:
test_loss, test_acurracy = model.evaluate(X_test, y_test)

1050/1 [================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

In [ ]:
print("Test accuracy: {}".format(test_acurracy))

Test accuracy: 0.4828571379184723


In [ ]:
test_loss

0.6946308978398641

# Explicações Finais - RNN

- No pré-processamento a variável nb_words estava com definição = 20000 a cerca do número de palavras, das quais são mais frequentes para se utilizar. Resolvi fazer um teste alterando o maxlenght para 100 mediante ao seu comprimento máximo de palavras.

- Notei que o modelo não estava se comportando muito bem, visto que mantive nb_words em 20000 e maxlenght em 500, das quais à accuracy encontrava-se entre 39% a 41%%, sem nenhum resultado satisfatório.

- Para ultrapassar os 41% em nossa análise, foi necessário usar 20 epochs com uma pequena alteração no lstm_units de 128 para 32. Assim, chegamos a 48% de accuracy.

Portando, diante dessa análise, ainda pretendo desafiar ainda mais accuracy para um resultado satisfatório, no qual o uso de CNN foi bem superior aos resultados usando uma RNN.